File: solutions/python/roman-numerals/1/roman_numerals.py
ROMANS = [("V", "I"), ("L", "X"), ("D", "C"), ("", "M"),]

def roman(number):
    if number > 3999 or number < 1:
        raise ValueError("Invalid number")
    out = ""
    for i, (five, one) in enumerate(ROMANS):
        last_digit = number % 10
        temp = ""
        if last_digit > 8:
            temp = one + ROMANS[i+1][1]
        elif last_digit > 5:
            temp = five + one * (last_digit - 5)
        elif last_digit == 5:
            temp = five
        elif last_digit == 4:
            temp = one + five
        else:
            temp = one * last_digit
        out = temp + out
        number //= 10
        if number == 0:
            break
    return out 
        

File: solutions/java/roman-numerals/1/src/main/java/RomanNumerals.java
class RomanNumerals {

    final int number;
    final String romanNumber;

    final static String [][] romanTable = new String[][] {
        {"V", "I"}, {"L", "X"}, {"D", "C"}, {"", "M"}
    };

    RomanNumerals(int number) {
        this.number = number;
        if (number < 1 || number > 3999) {
            throw new IllegalArgumentException("Invalid input");
        }
        StringBuilder sBuilder = new StringBuilder();
        for (int i = 0, n = romanTable.length; i < n; ++i) {
            String one = romanTable[i][1], five = romanTable[i][0];
            int lastDigit = number % 10;
            String appendNumber = switch(lastDigit) {
                case 9 -> one + romanTable[i + 1][1];
                case 8 -> five + one + one + one;
                case 7 -> five + one + one;
                case 6 -> five + one;
                case 5 -> five;
                case 4 -> one + five;
                case 3 -> one + one + one;
                case 2 -> one + one;
                case 1 -> one;
                default -> ""; // 0
            };
            sBuilder.insert(0, appendNumber);
            number /= 10;
            if (number == 0) break;
        }
        this.romanNumber = sBuilder.toString();
    }

    String getRomanNumeral() {
        return this.romanNumber;
    }
}
